File: pedRefiner.cpp
/*
 * Copyright (c) 2016 Hailin Su, ISU NBCEC
 *
 * This file is part of pedRefiner.
 *
 * pedRefiner is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * pedRefiner is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU General Lesser Public License
 * along with pedRefiner. If not, see <http://www.gnu.org/licenses/>.
 */

//
//  pedRefiner.cpp
//  libHailins
//
//  Created by Hailin SU on 4/15/14.
//

#include "pedRefiner.h"

using namespace std;
using namespace libcbk;

unsigned recGen(0), recGenMax(0);

// 20150629: prevent loop
string stem;
set<string> stemSet;
bool stemFault(false);

void usage(char* argv0)
{
    cout << endl;
    cout << "pedRefiner          by hailins@iastate.edu" << endl << endl;
    cout << "[Usage]" << endl << endl
    << argv0 << " anmList full_ped_csv_file [ -r | num_recursive_gen | xref-filename] > pedOut.csv 2>errLog" << endl << endl;
    cout << "[Description]" << endl
    << " - extracts the pedigree of all the ancestors of the animals in the anmList file;" << endl
    << " - if '-r' option is given, prints out all the descendants' IDs from the anmList file instead of print out the refined pedigree;" << endl
    << " - xref the pedigree info if specified a whitespace-delimited 3-col xref file with 1st col as command:" << endl
    << "       #          : line starting with sharp will be ignored " << endl
    << "       A ID1 ID2  : all ID1 (col1, 2 and 3) will be changed to ID2" << endl
    << "       S ID1 ID2  : all ID1 in the sire (2nd) column will be changed to ID2" << endl
    << "       D ID1 ID2  : all ID1 in the dam  (3rd) column will be changed to ID2" << endl
    << " - force accepting the latest version if multiple entries for the same animal were detected;" << endl
    << " - messages print out on screen (stderr, 2> to redirect)" << endl
    << " - results  print out on screen (stdout, 1> to redirect)" << endl
    << "            ancestors ordered prior than offsprings, delimetered by comma ',' " << endl << endl;
    cout << "[Updates]" << endl
    << " - version 2016-03-31 added -r option: will print out all the descendants' IDs from anmList, recursively to the end" << endl //
    << " - version 2015-07-13 added output: pedOut.ID2Gender according to the ped, and 'M' for unknown" << endl //
    << " - version 2015-07-02 added xref to correct animal IDs: replace 1st col with 2nd col" << endl //
    << " - version 2015-06-29 prevent segmentation fault 11 due to pedigree loop" << endl //
    << " - version 2015-06-12 added updating pre-loaded empty entries" << endl //
    << " - version 2015-04-15 added a 3rd parameter to determine the recursive generations" << endl //
    << " - version 2014-11-10 modified check() so that all the errors would be checked in one run" << endl
    << "                      added sort()" << endl
    << " - version 2014-04-15 initial version" << endl //
    << endl;
}

void reCout(const string &anm, PedMap &pm, set<string> &resultSet)
{
    if(resultSet.find(anm)!=resultSet.end())    // already in
        return;
    if(recGenMax>0 && ++recGen>recGenMax)
    {
        pm[anm]=Parents("0", "0");
        --recGen;
        return;
    }

    resultSet.insert(anm);
    if(pm.find(anm)!=pm.end())  // found
    {
        if(pm[anm].sire!="0")
            reCout(pm[anm].sire, pm, resultSet);
        if(pm[anm].dam!="0")
            reCout(pm[anm].dam, pm, resultSet);
        // 20150629: prevent loop
        if(pm[anm].dam==stem || pm[anm].sire==stem)
            stemFault=true;
    }
    else                        // not found
        pm[anm]=Parents("0", "0");
    --recGen;
}

void reSort(const string &anm, PedMap &pm, vector<string> &vecSortedID, set<string> &setSortedID)
{
    stemSet.insert(anm);
    if(setSortedID.find(anm)==setSortedID.end())    // if not in the sorted list
    {
        if(pm[anm].sire!="0" && !stemFault)                       // make sure parents get in the vec first
        {
            if(stemSet.find(pm[anm].sire)!=stemSet.end())   // loop found
            {
                cerr << endl << "         loop detected: [s] " << pm[anm].sire << endl;
                stemFault=true;
                vecSortedID.push_back(anm);
                setSortedID.insert(anm);
                return;
            }
            reSort(pm[anm].sire, pm, vecSortedID, setSortedID);
        }
        if(pm[anm].dam!="0" && !stemFault)
        {
            if(stemSet.find(pm[anm].dam)!=stemSet.end())
            {
                cerr << endl << "         loop detected: [d] " << pm[anm].dam << endl;
                stemFault=true;
                vecSortedID.push_back(anm);
                setSortedID.insert(anm);
                return;
            }
            reSort(pm[anm].dam, pm, vecSortedID, setSortedID);
        }
        vecSortedID.push_back(anm);
        setSortedID.insert(anm);
    }
    stemSet.erase(anm);
}

void reFillOffspring(const string &anm, PedMap &pm, set<string> &rs)
{
    set<string> setTmp;
    if(rs.find(anm)!=rs.end())  // if anm already in resultset, it means all its descendants are in the resultset
        return;

    if(recGenMax>0 && ++recGen>recGenMax)
    {
        --recGen;
        return;
    }
    if(pm.getOffspringSet(anm, setTmp))
    {
        for(set<string>::iterator sit=setTmp.begin(); sit!=setTmp.end(); ++sit)
        {
            reFillOffspring(*sit, pm, rs);
            rs.insert(*sit);
        }
        set<string>().swap(setTmp);
    }
    --recGen;
    rs.insert(anm);
}

int main (const int argc, char** argv)
{
    if((argc==5 && string(argv[3])!="-r") || (argc!=3 && argc!=4 && argc!=5))
    {
        cout << argc << " parameters read.";
        usage(argv[0]);
        exit (-1);
    }
    string lstFile(argv[1]), tgtFile(argv[2]);

    // read the list
    ifstream infs;
    infs.open(lstFile.c_str());
    if(!infs)
    {
        cerr << " * AnimalID list file " << lstFile.c_str() << " could not be open to read." << endl;
        exit (-2);
    }

    vector<string> vecAnmLst;
    string stmp;
    cerr << " - Loading AnimalID list" << endl;
    while(infs >> stmp)
        vecAnmLst.push_back(stmp);

    infs.close();

    stringstream sst;
    splitablestring aLine;
    PedMap pm;

    // 20160331: bR
    bool bR(false);
    if(argc==5)
    {
        bR=true;
        sst << argv[4];
        sst >> recGenMax;
        sst.clear();
    }
    else if(argc==4)
    {
        infs.open(argv[3]);
        if(!infs)   // not a file, treat it as "-r" or recursive generations parameter
        {
            if(string(argv[3])=="-r")
                bR=true;
            else
            {
                sst << argv[3];
                sst >> recGenMax;
                sst.clear();
            }
        }
        else        // a file, read the xref content
        {
            infs.close();
            cerr << " - loading xref" << endl;
            if(!pm.loadXrefMap(argv[3]))
            {
                cerr << "Xref not loaded. Aborting..." << endl;
                infs.close();
                exit(-3);
            }
        }
    }

    // pipe the target
    infs.open(tgtFile.c_str());
    if(!infs)
    {
        cerr << " * Target file " << tgtFile.c_str() << " could not be open to read." << endl;
        exit (-3);
    }

    cerr << " - Loading whole pedigree" << endl;
    while(getline(infs, aLine))
        pm.load(aLine, bR);
    infs.close();

    cerr << " - Checking pedigree for errors" << endl;
    pm.check();

    cerr << " - Filling output result set" << endl;
    set<string> resultSet;

    if(!bR) // 20160331: if it is a regular pedRefiner job
    {
        for(unsigned i=0; i<vecAnmLst.size(); ++i)
        {
            if(i%500==0)
                cerr << "\r          \r" << 100*i/vecAnmLst.size() << "%";
            // 20150629: prevent ped loop
            stem=vecAnmLst[i];
            recGen=1;
            reCout(stem, pm, resultSet);   // recursion
            if(stemFault)
                cerr << "\r              \r *L Pedigree loop detected while filling result set, stem = " << stem << endl;
            stemFault=false;
        }
        cerr << "\r              \r";

        cerr << " - Sorting" << endl;   // 2014-11-10
        vector<string>  vecSortedID;
        set<string>     setSortedID;
        unsigned ls(0);
        bool stemFaultStop(false);
        for(set<string>::iterator sit=resultSet.begin(); sit!=resultSet.end(); ++sit)
        {
            if(ls++ % 500 == 0)
                cerr << "\r       \r" << 100*ls/resultSet.size() << "%";
            stem=(*sit);
            reSort(stem, pm, vecSortedID, setSortedID);   // recursion
            if(stemFault)
            {
                cerr << " * Pedigree loop detected while sorting, stem = " << stem << endl;
                stemFaultStop=true;
            }
            stemFault=false;
            set<string>().swap(stemSet);
        }
        if(stemFaultStop)
        {
            cerr << endl
            << "####################################################################" << endl
            << "[1] print the output pedigree (with loop) to stderr" << endl
            << "[2] write the output pedigree (with loop) to file 'pedOut.wLoop.csv'" << endl
            << "[X] or other input: quit" << endl
            << "### Your choice (default=X): " << flush;
            string choice;
            cin >> choice;
            if(choice=="1")
            {
                for(ls=0; ls<vecSortedID.size(); ++ls)
                    cerr << "      " << pm.getRecord(vecSortedID[ls]) << endl;
            }
            else if(choice=="2")
            {
                ofstream ofs ("pedOut.wLoop.csv");
                if(ofs)
                {
                    for(ls=0; ls<vecSortedID.size(); ++ls)
                        ofs << "      " << pm.getRecord(vecSortedID[ls]) << endl;
                    ofs.close();
                }
            }
            exit (-3);
        }
        cerr << "\r       \r";
        cerr << " - Outputing" << endl;

        ofstream ofsID2Gender("pedOut.ID2Gender");
        for(ls=0; ls<vecSortedID.size(); ++ls)
        {
            if(ls % 500 == 0)
                cerr << "\r       \r" << 100*ls/vecSortedID.size() << "%";
            cout << pm.getRecord(vecSortedID[ls]) << endl;
            if(ofsID2Gender)
            {
                if(pm.mapID2Gender.find(vecSortedID[ls])!=pm.mapID2Gender.end())    // in list
                    ofsID2Gender << vecSortedID[ls] << " " << pm.mapID2Gender[vecSortedID[ls]] << endl;
                else                // use U as unknown
                    ofsID2Gender << vecSortedID[ls] << " U" << endl;
            }
        }
        cerr << "\r       \r";
        if(ofsID2Gender)
            ofsID2Gender.close();
    }
    else    // 20160331: if it is a rev pedRefiner job -r
    {
        // put lst in first
        const unsigned long sz(vecAnmLst.size());
        for(unsigned i=0; i<sz; ++i)
        {
            if(i%200==0)
                cerr << "\r                       \r  [ " << i+1 << " / " << sz << " ]";
            recGen=1;
            reFillOffspring(vecAnmLst[i], pm, resultSet);
        }
        cerr << "\r                       \r";

        cerr << " - Outputing" << endl;
        for(set<string>::iterator sit = resultSet.begin(); sit!=resultSet.end(); ++sit)
            cout << *sit << endl;
    }
}

File: pedRefiner.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
#  Copyright (c) 2017 Hailin Su, ISU NBCEC
#
#  This file is part of pedRefiner.
#
#  pedRefiner is free software: you can redistribute it and/or modify
#  it under the terms of the GNU Lesser General Public License as published
#  by the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  pedRefiner is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#  GNU Lesser General Public License for more details.
#
#  You should have received a copy of the GNU General Lesser Public License
#  along with pedRefiner. If not, see <http://www.gnu.org/licenses/>.
#
#
#  pedRefiner.cpp
#      Created by Hailin SU on 4/15/14.
#
#  pedRefiner.py
#      Created by Hailin SU on 1/23/17.

import os
import logging


class PedRefiner:
    def __init__(self):
        self.isValid = False
        self.rec_gen_max = 0
        self.rec_gen = 0
        self.missing_in = '.'      # I code it as '.' inside
        self.missing_out = '.'
        self.stem = ""
        self.stem_fault = False
        logging.basicConfig(level=logging.DEBUG)
        self.l = logging.getLogger('PedMap')
        # 20150702 mapXref
        self.mapXrefA = {}
        self.mapXrefS = {}
        self.mapXrefD = {}
        self.mapID2Gender = {}
        # 20150331 -r need mapID2Offspring
        self.mapID2SetOffspring = {}
        self.ped_map = {}     # holding original pedigree info
        self.opt_map = {}     # holding output pedigree set
        self.opt_vec = []
        self.l.info("PedRefiner().pipeline\n{}".format(self.pipeline.__doc__))

    def load_xref_map(self, file_name):
        """
        :param: file_name
        :return: bool, True if success
        """
        ret = False
        self.mapXrefA = {}
        self.mapXrefS = {}
        self.mapXrefD = {}
        if os.path.exists(file_name):
            ret = True
            with open(file_name, 'r') as fp:
                for line in fp:
                    vec_tmp = line.split()
                    if len(vec_tmp[0]) == 0 or vec_tmp[0][0] == "#":  # ignore empty or comment lines
                        continue
                    if len(vec_tmp) != 3:
                        self.l.error("PedMap::load_xref_map() ERROR - xref file {} is not a 3-col-through file."
                                     .format(file_name))
                        self.l.error("PedMap::load_xref_map() ERROR - line content [{}] contains {} fields."
                                     .format(line, len(vec_tmp)))
                        ret = False
                        break

                    if vec_tmp[0] == "A":
                        self.mapXrefA[vec_tmp[1]] = vec_tmp[2]
                    elif vec_tmp[0] == "S":
                        self.mapXrefS[vec_tmp[1]] = vec_tmp[2]
                    elif vec_tmp[0] == "D":
                        self.mapXrefD[vec_tmp[1]] = vec_tmp[2]
                    else:
                        self.l.warning('PedMap::loadXrefMap() WARNING - unknown command "{}", ignored'
                                       .format(line))

        return ret

    def load_ped(self, file_name, sep=",", missing_in='.', missing_out='.'):
        self.missing_in = missing_in
        self.missing_out = missing_out
        self.ped_map = {}
        if os.path.exists(file_name):
            with open(file_name, 'r') as fp:
                for line in fp:
                    self.isValid = self.__load_line(line, sep)
                    if not self.isValid:
                        break
        return self.isValid

    def __load_line(self, line, sep=","):
        ret = True
        vec_tmp = line.strip().split(sep)
        if len(vec_tmp) != 3:
            self.l.error(" * Error reading ped file: not a 3-col csv file, line starting with {}".format(vec_tmp[0]))
            self.l.error("   number of columns read: {}".format(len(vec_tmp)))
            ret = False
        else:
            for i in range(3):
                if len(vec_tmp[i]) == 0 or vec_tmp[i] == self.missing_in:
                    vec_tmp[i] = self.missing_out
                # 20150702 mapXref
                if vec_tmp[i] in self.mapXrefA:   # vec_tmp[i] is in the xrefA map
                    vec_tmp[i] = self.mapXrefA[vec_tmp[i]]
            idx, sire, dam = vec_tmp
            if sire in self.mapXrefS:       # vec_tmp[1] is in the xrefS map
                sire = self.mapXrefS[sire]
            if dam in self.mapXrefD:       # vec_tmp[2] is in the xrefD map
                dam = self.mapXrefD[dam]

            if idx != self.missing_out:
                p = (sire, dam)
                if idx in p:
                    self.l.warning(" #I ID {} is same for its parent(s), replaced with missing.".format(idx))
                    if idx == sire:
                        sire = self.missing_out
                    if idx == dam:
                        dam = self.missing_out
                    self.ped_map[idx] = (sire, dam)

                if idx not in self.ped_map:            # new entry
                    self.ped_map[idx] = p
                # duplicated but same record, just return: no-missing stored first
                # 20150729: partly canceled. overriding using new entry
                elif self.ped_map[idx][0] == sire and self.ped_map[idx][1] == dam:  # same
                    self.l.warning(" #D duplicated (but same) entry for {}".format(idx))

                # 20150612: override empty entry
                # 20150729: canceled. overriding using new entry
                else:
                    self.l.warning(" #D duplicated (yet quite different) entry for {}, force using version2"
                                   .format(idx))
                    self.l.warning("     version1: {0}{2}{1[0]}{2}{1[1]}".format(idx, self.ped_map[idx], sep))
                    self.ped_map[idx] = p
                    self.l.warning("     version2: {0}{2}{1[0]}{2}{1[1]}".format(idx, self.ped_map[idx], sep))

        return ret

    def check(self):
        ret = True
        if not self.isValid:
            self.l.warning(" * Errors occurred while reading pedigree.")
            ret = False

        map_s2i = {}
        map_d2i = {}   # map Sire/Dam 2 Integer(count)
        self.mapID2Gender = {}

        # filling the sets
        # self.l.debug("filling maps[ID -> parents]")
        for k in self.ped_map:
            s, d = self.ped_map[k]
            if s not in map_s2i:  # new
                map_s2i[s] = 1
                self.mapID2Gender[s] = "M"
            else:
                map_s2i[s] += 1

            if d not in map_d2i:  # new
                map_d2i[d] = 1
                self.mapID2Gender[d] = "F"
            else:
                map_d2i[d] += 1

        # make sure the two sets excludes each other
        # inserting intersection of Sire and Dam
        # print out the intersection and its count to support what gender it was
        intersection = [x for x in set(map_s2i.keys()).intersection(set(map_d2i.keys())) if x != self.missing_out]
        if len(intersection):
            self.l.warning(" *B Error: IDs appeared in both sire and dam columns. Use 'xref.CorrectB'" +
                           " as a xref file for the next run.")
            self.l.warning("   - creating 'xref.CorrectB'")
            ret = False
            with open('xref.CorrectB', 'w') as fp:
                for idx in intersection:
                    cnt_s = map_s2i[idx]
                    cnt_d = map_d2i[idx]
                    if cnt_s == cnt_d:
                        s_op = "A"
                    elif cnt_s > cnt_d:
                        s_op = "D"
                    else:
                        s_op = "S"

                    s_op = "{} {} {}\n".format(s_op, idx, self.missing_out)
                    fp.write(s_op)
        return ret

    def get_offspring_set(self, idx):
        ret = set()
        if idx in self.mapID2SetOffspring:
            ret = self.mapID2SetOffspring[idx]
        return ret

    # 20160331: a regular pedRefiner job
    def refine(self, list_file_name, opt_file_name, rec_gen_max=0, sep_out=",", flag_r=False):
        if os.path.exists(list_file_name):
            with open(list_file_name, 'r') as fp:
                anm_list = fp.read().splitlines()

            self.l.debug("we got {} individuals in the grep list".format(len(anm_list)))
            self.__populate_opt_map(anm_list, rec_gen_max)  # filling result set: opt_map
            self.l.debug("we got {} individuals in the output list".format(len(self.opt_vec)))

            if self.stem_fault:
                self.l.error(" *L Pedigree loop detected while filling result set, stem = {}".format(self.stem))
            self.isValid = True

            with open(opt_file_name, 'w') as fp:
                for idx in self.opt_vec:
                    fp.write("{0}{2}{1[0]}{2}{1[1]}\n".format(idx, self.opt_map[idx], sep_out))
        else:
            self.l.error(" * AnimalID list file {} could not be open to read.".format(list_file_name))
            self.isValid = False

        return self.isValid

    def __single_populate(self, idx):
        if idx in self.opt_map and self.rec_gen_max == 0:  # already in output ped, and no recGen restriction
            return

        self.rec_gen += 1
        if self.rec_gen > self.rec_gen_max > 0:
            if idx != self.missing_out:
                self.opt_map[idx] = (self.missing_out, self.missing_out)
                self.opt_vec.append(idx)
            self.rec_gen -= 1
            return

        if idx in self.ped_map:     # in input ped
            sire, dam = self.ped_map[idx]
            if sire != self.missing_out:
                self.__single_populate(sire)
            if dam != self.missing_out:
                self.__single_populate(dam)

            # 20150629: prevent loop
            if sire == self.stem or dam == self.stem:
                self.stem_fault = True
        else:  # not found
            sire = dam = self.missing_out

        if idx != self.missing_out:
            self.opt_map[idx] = (sire, dam)
            if idx not in self.opt_vec:
                self.opt_vec.append(idx)

        self.rec_gen -= 1

    def __populate_opt_map(self, anm_list, rec_gen_max=0):
        self.opt_map = {}
        self.opt_vec = []
        self.rec_gen_max = rec_gen_max
        for idx in anm_list:
            self.stem = idx
            self.rec_gen = 1
            self.__single_populate(idx)

    def pipeline(self, lst_fn, ped_fn, opt_fn, gen_max=0, flag_r=False, xref_fn=None, missing_in='.', missing_out='.',
                 sep_in=',', sep_out=','):
        """
        Suggested usage of PedRefiner: the pipeline() method.

        :param lst_fn:              file containing animal list to be grepped from the pedigree, one line each
        :param ped_fn:              input pedigree file, 3 col
        :param opt_fn:              output pedigree file, 3 col
        :param gen_max:     (0)     maximum recursive generation to grep for EVERYONE in lst_fn
        :param flag_r:      (False) bool, output descendant IDs if True
        :param xref_fn:     (None)  cross-reference file to modify output pedigree
        :param missing_in:  ('.')   missing value for input file, default = '.'
        :param missing_out: ('.')   missing value for output file, default = '.'
        :param sep_in:      (',')   separator for input file, default = ',' i.e. csv
        :param sep_out:     (',')   separator for output file, default = ',', i.e. csv


        Example

            from pedRefiner import PedRefiner
            pr = PedRefiner()
            pr.pipeline('animal_list', 'ped.input.csv', 'ped.output.csv', gen_max=3)

        """
        if xref_fn is not None:
            self.load_xref_map(xref_fn)
        if self.load_ped(ped_fn, sep_in, missing_in, missing_out) and self.check():
            self.refine(lst_fn, opt_fn, gen_max, sep_out, flag_r)

if __name__ == "__main__":
    pr = PedRefiner()
    pr.pipeline('/Volumes/data/epds/ASA_20150714_Carcass_analysis_w_BWCG/20161218-newRefine/lstID',  # lst_fn
                '/Volumes/data/epds/ASA_20150714_Carcass_analysis_w_BWCG/20161218-newRefine/pedAll.csv',
                '/Volumes/data/epds/ASA_20150714_Carcass_analysis_w_BWCG/20161218-newRefine/opt.pedpy.csv',
                3)
